client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/mongo/MongoMethodVisitor.java

```diff
@@ -0,0 +1,171 @@
+import java.util.Map;
+
+public class MongoMethodVisitor extends MethodVisitor {
+
+    private final String className;
+    private final String methodName;
+    private final String descriptor;
+
+    public MongoMethodVisitor(MethodVisitor mv,
+                              String className,
+                              String methodName,
+                              String descriptor) {
+        super(Constants.ASM, mv);
+
+        this.className = className;
+        this.methodName = methodName;
+        this.descriptor = descriptor;
+
+        fillMethodsToReplace();
+    }
+
+    /**
+     * Visits a method instruction. A method instruction is an instruction that invokes a method.
+     *
+     * @param opcode the opcode of the type instruction to be visited. This opcode is either INVOKEVIRTUAL, INVOKESPECIAL, INVOKESTATIC or INVOKEINTERFACE.
+     * @param owner  the internal name of the method's owner class (see getInternalName).
+     * @param name   the method's name.
+     * @param desc   the method's descriptor (see Type).
+     * @param itf    if the method's owner class is an interface.
+     */
+    @Override
+    public void visitMethodInsn(int opcode, String owner, String name,
+                                String desc, boolean itf) {
+
+        Method m = findMethodToReplace(owner, name, desc);
+        if (m == null) {
+            super.visitMethodInsn(opcode, owner, name, desc, itf);
+        } else {
+            replaceMethod(m);
+        }
+    }
+
+
+    /**
+     * A mapping owner->name->desc->method to replace existing methods
+     */
+    private final Map<String, Map<String, Map<String, Method>>> methodsToReplace = new HashMap<String, Map<String, Map<String, Method>>>();
+
+    /**
+     * Adds a method from method replacement such that matches the expected signature for
+     * a replacement method
+     *
+     * @param owner  the owner of the invoke
+     * @param name   the name of the method being called
+     * @param desc   the signature of the method being called
+     * @param method the new method that will replace the original one
+     */
+    private void addMethodToReplace(String owner, String name, String desc, Method method) {
+        if (!methodsToReplace.containsKey(owner)) {
+            methodsToReplace.put(owner, new HashMap<>());
+        }
+        if (!methodsToReplace.get(owner).containsKey(name)) {
+            methodsToReplace.get(owner).put(name, new HashMap<>());
+        }
+        methodsToReplace.get(owner).get(name).put(desc, method);
+    }
+
+
+    private void fillMethodsToReplace() {
+
+        try {
+
+            Class findIterableClass = Class.forName("com.mongodb.client.FindIterable");
+            Class bsonClass = Class.forName("org.bson.conversions.Bson");
+            Class mongoCollectionClass = Class.forName("com.mongodb.client.MongoCollection");
+            Class clientSessionClass = Class.forName("com.mongodb.client.ClientSession");
+
```

Comment:
I am working on a generic way to handle third-party dependencies in method replacements. so hopefully this can be simplified :)

Reply:
ok